### PR TITLE
[DEV-001] Formalize Schema

### LINKED ISSUE #1: Formalize Schema

### PR DESCRIPTION
**Pull request by nivykani**
## [DEV-001] Formalize Schema

Fixes: #1 

-   [x] Has the code been formatted?
-   [x] Have dependencies/packages been added? (N/A)

### Changes
Tables for the following:
- users: all users
- institutions: universities or independent departments
    - semesters: groupings of courses (e.g. "Fall 2020")
    - administrators: admins per institution
    - enrolment: list of courses a student is enrolled in, and their position (e.g. student) in each
    - courses: names and course codes
        - categories: list of tags available for a post (e.g. "pinned", "assignment 1")
        - post_categories: shows all the tags on each post
        - thread: main posts and top-level replies (refer to mock-ups)
        - comments: bottom-level replies (text-only)
            - polls: max four-option text polls that are associated with a thread and have an end date
            - poll-results: stores user_id/poll_id/choice, allows multiple votes
        - award_types: list of possible awards 
        - awards: list of awards given to posts

### Testing
-   [x] Does the code have sufficient automated test coverage? If not, include a reasoning and steps required to manually QA (N/A)

### CODE REVIEW

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]: Instead of ALTER TABLE to add the foreign key, you can use
- a syntax like for categories `course REFERENCES courses(id) ON DELETE CASCADE`

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]: instead of ID, can have this be more specific, e.g., userID. If we use just id, it can be kind of difficult to read.

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]: semester should be related to the institution, different institutions use different terminologies :)

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]: If you define them as a PRIMARY KEY, the value cannot be null IIRC

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]: Are we allowing nested comments? What would a query look like for all posts, it might get very difficult to reason about queries on this

**Review by krishchow — CHANGES_REQUESTED** [review]

File: backend/storage/discussion-board.sql
Comment (nivykani) [reply backend/storage/discussion-board.sql]: Changed to "semester_name" (so semesters aren't tied to a season/year) and added a reference to the institution

File: backend/storage/discussion-board.sql
Comment (jcserv) [inline backend/storage/discussion-board.sql]: should we call it post or perhaps thread?
thread conveys that its for discussion moreso than post

File: backend/storage/discussion-board.sql
Comment (nivykani) [reply backend/storage/discussion-board.sql]: For the second question, now each post references its course, since a post shouldn't be about more than one course (because posts belong to categories which are unique to courses). So it should be straightforward to find all posts under a course, even without filtering by category.

For the first question, a separate "replies" table makes sense since top-level posts have some extra fields... but then the parent_id could be either a post or a reply... or maybe just limit the nesting of comments on the front end?

File: backend/storage/discussion-board.sql
Comment (jcserv) [inline backend/storage/discussion-board.sql:69]: taking a look at the backlog, i think we could add a few fields for future features such as:
- is_locked (indicates whether a thread is no longer taking replies)
- is_anon (indicates whether we should hide the poster's identity)

some other considerations (that im not quite sure of but we should definitely discuss!):
- how will we handle a thread being "pinned"?
- will polls be considered a post/thread or require a different table?

File: backend/storage/discussion-board.sql
Comment (jcserv) [inline backend/storage/discussion-board.sql:5]: one nice thing about piazza is that it keeps track of contributions!
we should probably have a field for this

File: backend/storage/discussion-board.sql
Comment (nivykani) [reply backend/storage/discussion-board.sql:69]: for 1), when a new course is added, we could always add a category with category_name "pinned" under that course? Threads with the "pinned" category can be displayed first.

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  institution_id SERIAL PRIMARY KEY,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql:16]: This is more meta, but what is `location`, is it a string representation of the coordinates? an address?

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  semester_idSERIAL PRIMARY KEY,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  course_id SERIAL PRIMARY KEY,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  question_contributions int DEFAULT 0 NOT NULL,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  answer_contributions int DEFAULT 0 NOT NULL,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  category_id SERIAL PRIMARY KEY,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  course_id int REFERENCES courses (course_id ) ON DELETE CASCADE
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  thread_id SERIAL PRIMARY KEY,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  parent_id int REFERENCES thread (thread_id) ON DELETE CASCADE,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  num_children int DEFAULT 0 NOT NULL,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  course_id int REFERENCES courses (course_id) ON DELETE CASCADE,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  body text NOT NULL,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  upvotes int DEFAULT 0 NOT NULL,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  view_count int DEFAULT 0 NOT NULL,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [reply backend/storage/discussion-board.sql:69]: similar to is_anon, how are we determining anonymous names, I assume like Piazza we might want to generate it once per thread, if so, are we doing this deterministically (so we don't need to store any additional information), or are we doing it randomly, and we do need to store some attribute in the thread/comment?

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  award_type_id SERIAL PRIMARY KEY,
```

File: backend/storage/discussion-board.sql
Comment (krishchow) [inline backend/storage/discussion-board.sql]:
```suggestion
  award_type int REFERENCES award_types (award_type_id) ON DELETE CASCADE,
```

**Review by krishchow — COMMENTED** [review]
Lot's of minor nitpicks, but the general idea is that we want to utilize the database for as much of our validation as possible. The idea here is that we can make guarantees about the data that we receive in the end.

File: backend/storage/discussion-board.sql
Comment (krishchow) [reply backend/storage/discussion-board.sql]: If we want to try and find a happy medium between SO and piazza, we might want to differentiate between replies to a post, and the comments below that reply. The comments can be much simpler, just like SO comments, what do you think about this idea?

File: backend/storage/discussion-board.sql
Comment (nivykani) [reply backend/storage/discussion-board.sql:16]: I think I imagined it like a drop-down menu of major cities, or "other" (custom) if their city isn't listed? (though the location fields would be empty until we decide what we want to do with the locations, bc some universities span across cities.)

File: backend/storage/discussion-board.sql
Comment (nivykani) [reply backend/storage/discussion-board.sql]: Yup, makes sense to differentiate between posts/replies and comments! Added a "comments" table

Comment (nivykani) [reply backend/storage/discussion-board.sql]: Since posts and replies (not comments) have similar fields except title and view_count, they can be in the same table right? That way, awards can be given to both posts and replies.

File: backend/storage/discussion-board.sql
Comment (nivykani) [reply backend/storage/discussion-board.sql:69]: the latest commit has a separate table for polls, where the polls reference a post (or a top-level reply). The poll's votes are stored in yet another table, but I think this can allow for polls where people can click on multiple options

**Review by jcserv — APPROVED** [review]
Great progress Nivy! You've taken all of our feedback and implemented it thoughtfully which is great. LGTM ✔️ 
Wait until Krish's approval to merge in though!

Some feedback for the PR process:
- Try not to over-commit! There are 28 commits for this PR which is a lot!
- The PR template was not filled out correctly (title, checkboxes)

**Comment by krishchow** [comment]
[Proposed changes](https://i.imgur.com/r58PNma.png)

The idea here is to change the schema to achieve the following goals

- Be more flexible to work with different types of threads
- Allow users to vote, and unvote certain option
- Allows admins to easily change/delete poll options

Not a complete diagram, but that's the core idea

What do you think? @jcserv @NivedithaK

**Review by krishchow — APPROVED** [review]
LGTM

**Pull request merged**
